internal/handlers/common/saslstart_plain.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+)
+
+// saslStartPlain extracts username and password from PLAIN `saslStart` payload.
+func saslStartPlain(doc *types.Document) (string, string, error) {
+	var payload []byte
+
+	// some drivers send payload as a string
+	stringPayload, err := GetRequiredParam[string](doc, "payload")
+	if err == nil {
+		if payload, err = base64.StdEncoding.DecodeString(stringPayload); err != nil {
+			return "", "", commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrBadValue,
+				fmt.Sprintf("Invalid payload: %v", err),
+				"payload",
+			)
+		}
+	}
+
+	// most drivers follow spec and send payload as a binary
+	binaryPayload, err := GetRequiredParam[types.Binary](doc, "payload")
+	if err == nil {
+		payload = binaryPayload.B
+	}
+
+	// if not string nor binary payload was sent, return a validation error
```

Comment:
what validation error?

> // return error about expected types.Binary, not string

I think that comment was important – it explains why we check less common thing first

Reply:
hehe :) the initial comment doesn't explain to me why, that's why I removed it.
it explains what is done, but not why. so, why we can't process binary first?

Reply:
Because we re-assign the `err` variable, and we want to return a errors messages about expecting binary payload, not string payload – just like the removed comment says

Reply:
but why? why an error about the string would not fit? what's the difference?

Reply:
… because the binary payload is more common and used by most drivers. All that is already written above. Please give it 5 minutes.

Reply:
so what? do we want it for compatibility or something? we can return an error like `[binary, string]` like we do in other places. why this one is different?